polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -2101,6 +2098,69 @@ MLIRScanner::EmitGPUCallExpr(clang::CallExpr *expr) {
   return std::make_pair(ValueCategory(), false);
 }
 
+static NamedAttrList getSYCLMethodOpAttrs(OpBuilder &builder,
+                                          mlir::Type funcType,
+                                          llvm::StringRef funcTypeName,
+                                          llvm::StringRef name,
+                                          llvm::StringRef mangledName) {
+  NamedAttrList attrs;
+  attrs.set(mlir::sycl::SYCLDialect::getBaseTypeAttrName(),
+            mlir::TypeAttr::get(funcType));
+  attrs.set(mlir::sycl::SYCLDialect::getBaseTypeNameAttrName(),
+            FlatSymbolRefAttr::get(builder.getStringAttr(funcTypeName)));
+  attrs.set(mlir::sycl::SYCLDialect::getFunctionNameAttrName(),
+            FlatSymbolRefAttr::get(builder.getStringAttr(name)));
+  attrs.set(mlir::sycl::SYCLDialect::getMangledFunctionNameAttrName(),
+            FlatSymbolRefAttr::get(builder.getStringAttr(mangledName)));
+  return attrs;
+}
+
+llvm::Optional<SYCLMethodOpInterface>
+MLIRScanner::createSYCLMethodOp(llvm::StringRef name, mlir::ValueRange operands,
+                                mlir::TypeRange types, llvm::StringRef funcType,
+                                llvm::StringRef mangledName) {
+  // Not a method call.
+  if (operands.empty() || !operands[0].getType().isa<MemRefType>()) {
+    return llvm::None;
+  }
+
+  auto *SYCLDialect =
+      operands[0].getContext()->getLoadedDialect<mlir::sycl::SYCLDialect>();
+  assert(SYCLDialect && "MLIR-SYCL dialect not loaded.");
+
+  // Need to copy to avoid overriding elements in the input argument.
+  SmallVector<mlir::Value> operandsCpy(operands);
+
+  // SYCLCastOps are abstracted to avoid missing method calls due to
+  // implementation details.
+  if (auto Cast = operandsCpy[0].getDefiningOp<mlir::sycl::SYCLCastOp>()) {
+    operandsCpy[0] = Cast->getOperand(0);
```

Comment:
The cast is removed in order to abstract implementation-defined types, i.e., so that methods such as the `range::get` can be represented as an operation receiving a reference to a `range`, not to the underlying `array` type. It is later reintroduced in the LLVM lowering pass if needed.

Reply:
This is dangerous and makes me question whether we can find an alternative design. I'm concerned about 2 things:
   - removing the cast makes is akin to hiding it inside the operation overloading the semantic of the operation
   - adding the cast back is necessary to recover the correct semantics of the program and seems like work that was created only because of a design decision

Suppose we have 2 get member functions, one in a base class and one in the derived class:
```
struct A { int get(); };
struct B : public A { int get(); };

int foo() {
  B *pb = new B();
  return ((A *)pb)->get() + pb->get();
}
```
A call to `B::get` would receive a `this` pointer with type `B*`, and a call to `A::get` would need to receive a `this` pointer with type `A*`.  If both calls are converted into the same MLIR operation then the operation would need a way to differentiate between the 2 scenarios above. 

I'd like to make sure we can handle that scenario with the proposed design.

Reply:
However, this does not interfere with the work on the other PR, so, if you're OK with this, I will work on that tomorrow so that, hopefully, we can close this before the weekend